File: decorator/python/data_source.py
from abc import ABC, abstractmethod


class DataSource(ABC):
    """
    Üst sınıfların uygulayacağı arayüzdür.
    Yapılacak temel işlemler tanımlanmıştır.
    UML diyagramındaki Component arayüzüne denk gelmektedir.
    """

    @abstractmethod
    def get_file_name(self):
        pass

    @abstractmethod
    def write_data(self, data):
        pass

    @abstractmethod
    def read_data(self):
        pass

File: decorator/go/data-source.go
package main

// DataSource arayüzünde yapılacak temel işlemler tanımlanmıştır.
// UML diyagramındaki Component arayüzüne denk gelmektedir.
type DataSource interface {
	GetFileName() string
	WriteData(data []byte)
	ReadData()
}
